### PR TITLE
[SPARK-34795][SQL][TESTS] Adds a new job in GitHub Actions to check the output of TPC-DS queries

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR proposes to add a new job in GitHub Actions to check the output of TPC-DS queries.

NOTE: I've checked that the new job took 17m 35s in the GitHub Actions env.

### Why are the changes needed?

There are some cases where we noticed runtime-realted bugs after merging commits (e.g. .SPARK-33822). Therefore, I think it is worth adding a new job in GitHub Actions to check query output of TPC-DS (sf=1).

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

The new test added.

### CODE REVIEW

File: .github/workflows/build_and_test.yml
Comment (maropu) [inline .github/workflows/build_and_test.yml:443]: Some files exceed the maximum single file size (100mb) defined by GitHub, so  I used `lfs` to upload them.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestSuite.scala:385]: I just removed the unnecessary code here.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:67]: I modified the code in `isSorted` because we still need to sort output rows even if `plan` has a `Sort` node. For example, most sort nodes sort rows by some columns of all output columns, but we always need to sort them by  all output columns. This change leads to many diffs of the files in `sql/core/src/test/resources/sql-tests/results`.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQueryTestHelper.scala:67]: The other code parts in this trait are just copied from `SQLQueryTestSuite`.

**Comment by maropu** [comment]
NOTE: I saw the Xiao's comment in [the recent release process](http://apache-spark-developers-list.1001551.n3.nabble.com/VOTE-Release-Spark-3-1-1-RC3-tp30788p30830.html) and then I made this PR.

File: .github/workflows/build_and_test.yml
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:443]: To reduce the network traffic, shall we use more storage-efficient format like gzip or zstd? For example, the following is `zstd` result.
```
$ aws s3 ls s3://ttl-35d-813987666268/users/dongjoon/tpcds/1g-zstd/ --recursive --human-readable | grep MiB | sort -k3
2021-03-12 14:06:32    1.0 MiB users/dongjoon/tpcds/1g-zstd/item/part-00000-cc566e52-0113-4bb3-9fe5-ba69de177fc2-c000.zstd.parquet
2021-03-12 14:04:43    3.6 MiB users/dongjoon/tpcds/1g-zstd/customer/part-00000-068120d1-27e5-4553-8d2d-e75143aa1da4-c000.zstd.parquet
2021-03-12 14:03:49    4.2 MiB users/dongjoon/tpcds/1g-zstd/web_returns/part-00000-f5ef2333-1d55-4655-9e4f-55072eb88e41-c000.zstd.parquet
2021-03-12 14:05:19    4.9 MiB users/dongjoon/tpcds/1g-zstd/customer_demographics/part-00000-279d7119-dc0f-4267-a190-977a630c5443-c000.zstd.parquet
2021-03-12 14:00:48    7.8 MiB users/dongjoon/tpcds/1g-zstd/catalog_returns/part-00000-eefacec3-ac03-4202-bbaf-7fc9d5caa11a-c000.zstd.parquet
2021-03-12 14:02:51   12.1 MiB users/dongjoon/tpcds/1g-zstd/store_returns/part-00000-bb811271-b6c3-4e33-be26-4208c7d6fe7a-c000.zstd.parquet
2021-03-12 14:01:13   14.9 MiB users/dongjoon/tpcds/1g-zstd/inventory/part-00000-27cbda33-87d3-4f2a-8df4-2c2fd63b420c-c000.zstd.parquet
2021-03-12 14:03:19   34.4 MiB users/dongjoon/tpcds/1g-zstd/web_sales/part-00000-b9dff830-93f2-4514-ab2f-4cb3f4c7a704-c000.zstd.parquet
2021-03-12 14:00:08   72.6 MiB users/dongjoon/tpcds/1g-zstd/catalog_sales/part-00000-7b3d90bc-1bd3-4396-a039-c11b2fad45fa-c000.zstd.parquet
2021-03-12 14:01:54   91.4 MiB users/dongjoon/tpcds/1g-zstd/store_sales/part-00000-553b44ad-8032-4b0f-a34b-f72c41d17787-c000.zstd.parquet
```

**Comment by dongjoon-hyun** [comment]
cc @wangyum , @HyukjinKwon because @wangyum made this attempt already before.
- https://github.com/apache/spark/pull/31303

File: .github/workflows/build_and_test.yml
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:443]: Since we are still not using Apache Parquet 1.12.0, `gzip` will be the only available option here.

**Comment by maropu** [comment]
Yea, but this PR is not for the benchmark purpose.

File: .github/workflows/build_and_test.yml
Comment (maropu) [reply .github/workflows/build_and_test.yml:443]: Ah, I see. Nice suggestion. I'll follow it.

**Comment by HyukjinKwon** [comment]
I prefer to have this in GA ... cc @wangyum who should be the best person to review.
cc @cloud-fan too

File: .github/workflows/build_and_test.yml
Comment (wangyum) [reply .github/workflows/build_and_test.yml:443]: Another option is we can generate data at runtime: https://github.com/apache/spark/pull/31303

**Comment by cloud-fan** [comment]
+1 to have this check!

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q1.sql.out
Comment (wangyum) [inline sql/core/src/test/resources/tpcds-query-results/v1_4/q1.sql.out:22]: Could we do not generate the query to reduce the file size?

Comment (maropu) [reply sql/core/src/test/resources/tpcds-query-results/v1_4/q1.sql.out:22]: okay.

**Comment by maropu** [comment]
NOTE: I'm checking if the result in the golden files is correct query-by-query.

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q1.sql.out
Comment (maropu) [reply sql/core/src/test/resources/tpcds-query-results/v1_4/q1.sql.out]: done

File: .github/workflows/build_and_test.yml
Comment (maropu) [reply .github/workflows/build_and_test.yml]: okay, I've changed the format from snappy to gzip.

Comment (maropu) [reply .github/workflows/build_and_test.yml]: > Another option is we can generate data at runtime: #31303

Yea, we can, but I think we don't need to re-generate these data per each invoke.

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q17.sql.out
Comment (maropu) [inline sql/core/src/test/resources/tpcds-query-results/v1_4/q17.sql.out:6]: This should be 
```
+AAAAAAAAKPFEAAAA       Recently right  TN      1       99.0    NULL    NULL    1       66.0    NULL    NULL    1       32.0    NULL    NULL
```
A PR to fix this bug is: https://github.com/apache/spark/pull/31940

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q41.sql.out
Comment (maropu) [inline sql/core/src/test/resources/tpcds-query-results/v1_4/q41.sql.out:6]: This should be:
```
ableationableought                                
anticallyeingese                                  
callycallyeingese                                 
oughtationableought                               
```
A PR to fix this is: https://github.com/apache/spark/pull/31940

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q17.sql.out
Comment (maropu) [reply sql/core/src/test/resources/tpcds-query-results/v1_4/q17.sql.out:6]: This seems to be caused by the wrong schema of `date_dim`: https://github.com/apache/spark/pull/31012/files#r599447123

**Review by dongjoon-hyun — DISMISSED** [review]
This PR seems to have two independent themes. Could you split this PR into two?
- One is having `build_and_test.yml` change only.
- One is the other test suite change and golden-file fix.

**Comment by maropu** [comment]
okay, @dongjoon-hyun ~ After I finish checking all the TPCDS query results, I'll do that (I'm still checking them)

**Comment by maropu** [comment]
Split off the unrelated part from this PR: https://github.com/apache/spark/pull/31946

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q78.sql.out
Comment (maropu) [inline sql/core/src/test/resources/tpcds-query-results/v1_4/q78.sql.out:6]: This query returns nothing when `sf=1`. I've checked this query returns output rows if `sf` increases, but I think larger data sets can make a hight pressure on GA jobs, so I keep this test as it is just for output schema checks.

File: sql/core/src/test/resources/tpcds-query-results/v2_7/q78.sql.out
Comment (maropu) [inline sql/core/src/test/resources/tpcds-query-results/v2_7/q78.sql.out:6]: The same comment with https://github.com/apache/spark/pull/31886/files#r600489102

File: sql/core/src/test/resources/tpcds-query-results/v2_7/q24.sql.out
Comment (maropu) [inline sql/core/src/test/resources/tpcds-query-results/v2_7/q24.sql.out:6]: The same comment: https://github.com/apache/spark/pull/31886/files#r600489102

File: sql/core/src/test/resources/tpcds-query-results/v1_4/q17.sql.out
Comment (maropu) [reply sql/core/src/test/resources/tpcds-query-results/v1_4/q17.sql.out]: Fixed in #31012

**Comment by maropu** [comment]
okay, ready to review. cc: @dongjoon-hyun @HyukjinKwon @wangyum

**Comment by HyukjinKwon** [comment]
I support this one but it would be great if this gets some reviews from @wangyum.

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:436]: Maybe we should do the caches here as we do above "Cache Coursier local repository" since it builds and runs the test with downloading new dependencies in any event.

File: .github/workflows/build_and_test.yml
Comment (wangyum) [inline .github/workflows/build_and_test.yml:442]: @dongjoon-hyun Is it acceptable to use a personal repository?

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:442]: Either https://github.com/maropu/spark-tpcds-datagen or https://github.com/maropu/spark-tpcds-datagen doesn't look having a license problem. I don't see any problem of using it because we're already using other repos in our CI. GA plugins, or other personal repos such as https://github.com/apache/spark/blob/master/dev/appveyor-install-dependencies.ps1#L98

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:454]: I would name it "Run TPC-DS"

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:441]: Can we follow https://github.com/apache/spark/pull/31303/files#diff-48c0ee97c53013d18d6bbae44648f7fab9af2e0bf5b0dc1ca761e18ec5c478f2R441-R450? I would be great if we can just stick to using the original repo with addressing https://github.com/apache/spark/pull/31303/files#r563236578. cc @wangyum FYI.

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala:156]:
```suggestion
    ignore("skipped because env `SPARK_TPCDS_DATA` is not set") {}
```

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala:56]: no bit deal but maybe we could use `sys.env.get("SPARK_TPCDS_DATA")` and use `None` to express that it's not set.

File: sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/TPCDSQueryTestSuite.scala:57]: and maybe: `sys.env.get("SPARK_GENERATE_GOLDEN_FILES").exists(_ == "1")`

File: .github/workflows/build_and_test.yml
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:442]: Thank you for pinging me, @wangyum . Ya, I hope we can bring this into Apache Spark eventually.  For now, let's move on. I also need to move GitHub Action linux image into Apache Spark repo in this year.

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:442]: Thank you @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
I didn't check the result one-by-one. Feel free to proceed to merge when you think this PR is enough, @maropu , @HyukjinKwon , @wangyum .

File: .github/workflows/build_and_test.yml
Comment (maropu) [reply .github/workflows/build_and_test.yml]: Yea, I also think it is the best to move the repo into the official one finally.

File: .github/workflows/build_and_test.yml
Comment (maropu) [reply .github/workflows/build_and_test.yml]: How about `Run TPC-DS queries`?

File: .github/workflows/build_and_test.yml
Comment (maropu) [reply .github/workflows/build_and_test.yml]: Ah, I see. To generate table data for `TPCDSQueryTestSuite`, I think we need more fixes  in the orignal repo because they have different TPC-DS schemas (char/varchar vs string):
spark-sql-perf: https://github.com/databricks/spark-sql-perf/blob/master/src/main/scala/com/databricks/spark/sql/perf/tpcds/TPCDSTables.scala#L57-L542
spark-master/branch-3.1: https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/TPCDSBase.scala#L51-L548

I've filed a ticket for this: https://github.com/databricks/spark-sql-perf/issues/198 I'll make a PR for it when I have time later.

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:450]: Thanks @maropu. Just to clarify, do you need https://github.com/databricks/spark-sql-perf/pull/196 too?

**Comment by wangyum** [comment]
Please wait me a few hours. I will verify the result first.

**Comment by HyukjinKwon** [comment]
@wangyum, don't worry. There are few more issues to address (https://github.com/apache/spark/pull/31886#discussion_r603307756). It will take more.

File: .github/workflows/build_and_test.yml
Comment (maropu) [reply .github/workflows/build_and_test.yml:450]: yea, the @wangyum PR looks useful when generating data. (that PR is not a blocker of this PR though)

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:450]: Sure, I will ping him offline to review

Comment (maropu) [reply .github/workflows/build_and_test.yml:450]: Thanks, @HyukjinKwon ~

**Review by HyukjinKwon — APPROVED** [review]
I think this is fine. I prefer to switch `maropu/spark-tpcds-sf-1` to `databricks/spark-sql-perf`. I roughly checked it internally, and it seems fine to open source this repo in the future as well.

We could merge this one, and maybe @wangyum can follow-up to switch to `databricks/spark-sql-perf`? I will leave it to both of you to merge and how to proceed.

LGTM (pending https://github.com/apache/spark/pull/31886#issuecomment-809927983) in any event.

**Review by wangyum — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @HyukjinKwon @wangyum . Merged to master.

**Comment by maropu** [comment]
If someone finds the failure of the new GA job, please let me know. cc: @dongjoon-hyun @cloud-fan @viirya @gatorsmile

**Comment by maropu** [comment]
After we resolve the pending issues on databricks/spark-sql-perf, I'm planning to add this job into branch-3.1./3.0, too.

**Comment by HyukjinKwon** [comment]
Thanks @maropu.

**Comment by viirya** [comment]
Thanks @maropu

**Comment by gatorsmile** [comment]
This is awesome! We should do it 5 years ago. :-)

**Pull request closed without merging**
